### PR TITLE
DATAUP-468 cleanup part 2

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes

Mainly migrate the `isEqual` function from the `unodep` module to the `lang` module (a deep equality checker's kind of a language extension right?) and tests that module.

I also refactored the `pRequire` function, which wraps a require call in a Promise. Very similar code was also copy-pasted in a few different places. More commentary inline, but to be honest, I'm not sure what this function buys us in some places over just importing the modules as dependencies (aside from ones where the module name is chosen on the fly), but I haven't changed those parts.

# Jira Ticket / Issue #
https://kbase-jira.atlassian.net/browse/DATAUP-468
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
Again, should be no visible changes.
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/buildTools/loadAppWidgets.js [hunk 1]
```diff
@@ -99,7 +97,6 @@ define([
     'common/spec',
     'common/specValidation',
     'common/ui',
-    'common/unodep',
```
Comment (briehl) [inline kbase-extension/static/buildTools/loadAppWidgets.js:102]: goodbye, `unodep`

File: kbase-extension/static/kbase/js/common/fsm.js [hunk 2]
```diff
@@ -5,7 +5,7 @@
  *
  */
 
-define(['./unodep', 'common/runtime'], (utils, Runtime) => {
+define(['./lang', 'common/runtime'], (utils, Runtime) => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/common/fsm.js:8]: less `unodep`, more `lang`

File: kbase-extension/static/kbase/js/common/lang.js [hunk 3]
```diff
@@ -7,18 +7,80 @@ define(['bluebird'], (Promise) => {
...
-                resolve(arguments);
+            require(modules, (...args) => {
+                resolve(args);
             }, (err) => {
                 reject(err);
             });
         });
     }
 
+    function isEqual(v1, v2) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/common/lang.js:26]: copied, and slightly adjusted to remove unreachable if-branches, from `unodep`

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/customInput.js [hunk 4]
```diff
@@ -1,39 +1,23 @@
 define([
     'bluebird',
-    'require',
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/customInput.js:3]: Adjusted some things in here while converting its own `pRequire` -> `lang.pRequire`. Mostly either linting or deleting unused code/dependencies.

File: nbextensions/appCell2/widgets/appCellWidget.js [hunk 5]
```diff
@@ -157,237 +157,235 @@ define([
...
         function loadParamsWidget(arg) {
-            return pRequire(['./appParamsWidget']).spread((Widget) => {
-                // TODO: widget should make own bus.
-                const widgetBus = runtime
-                        .bus()
-                        .makeChannelBus({ description: 'Parent comm bus for input widget' }),
-                    widget = Widget.make({
-                        bus: widgetBus,
-                        initialParams: model.getItem('params'),
+            return lang
```
Comment (briehl) [inline nbextensions/appCell2/widgets/appCellWidget.js:161]: This is almost entirely Prettier trying its hardest throughout the file.

File: nbextensions/appCell2/widgets/appCellWidget.js [hunk 6]
```diff
@@ -157,237 +157,235 @@ define([
...
+                .pRequire(['nbextensions/appCell2/widgets/appParamsWidget'])
```
Comment (briehl) [inline nbextensions/appCell2/widgets/appCellWidget.js:162]: This is the main change that was made - use `lang.pRequire` instead of its own local copy. This is one of the spots that makes me scratch my head - if we already know the name and path to the module to load, why does it need to be loaded later rather than as a top level dependency? It doesn't really execute anything, and it's not like it's multiple MB of code. 🤷

File: test/unit/spec/common/lang-Spec.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
+define(['common/lang'], (Utils) => {
```
Comment (briehl) [inline test/unit/spec/common/lang-Spec.js:1]: Testing the lang module all through. Some edge cases in here that could maybe be handled better, but it should all be readable enough.

File: nbextensions/appCell2/widgets/appCellWidget.js (hunk 6 above)
Comment (ialarmedalien) [reply nbextensions/appCell2/widgets/appCellWidget.js:162]: I've also scratched my head around this area, as well as doing a bit of

head-desk-interaction

If we could shift over to using a similar pattern for the params edit/view as in the bulk import cell, some of this could just disappear.

File: kbase-extension/static/kbase/js/common/lang.js (hunk 3 above)
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/common/lang.js:26]: we have underscore as one of the deps -- maybe we should just use the underscore .isEqual function?

File: test/unit/spec/common/lang-Spec.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                const obj = {
+                    foo: {
+                        bar: 'baz',
+                    },
+                };
+
+                const copied = Utils.copy(obj);
+                expect(copied).toEqual(obj);
+                copied.foo.bar = 'wat';
+                expect(copied).not.toEqual(obj);
```
Comment (ialarmedalien) [inline test/unit/spec/common/lang-Spec.js:38]: could you put in an explicit test:
```js
expect(obj.foo.bar).toEqual('baz');
```

Just to ensure that it's all working as expected

Comment (briehl) [reply test/unit/spec/common/lang-Spec.js:38]: Ok, done.

File: kbase-extension/static/kbase/js/common/lang.js (hunk 3 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/common/lang.js:26]: Yeah, that sounds reasonable. I didn't think of that.
We should update to the latest underscore, too.

**Comment by briehl** [comment]
From this conversation and other thoughts, I think it's worth doing the following:
1. Remove the pRequire call from the few places where it imports a statically named widget, and just import that widget in the top-level define statement (I think that's two or three modules).
2. Convert the isEqual call to the same call in underscore, which is already a dependency.
3. Bump underscore to the latest version (1.13.1, released in April this year).
4. Remove `isEqual` from `lang.js` entirely.

**Comment by briehl** [comment]
@ialarmedalien Updates made, ready for re-review!

File: nbextensions/bulkImportCell/bulkImportCell.js [hunk 9]
```diff
@@ -42,7 +43,6 @@ define([
     UI,
     Utils,
     BulkImportUtil,
-    SimpleUtil,
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/bulkImportCell.js:45]: another name!

**Review by ialarmedalien — APPROVED** [review]
🎉
